src/StateHistoryPostgresActionReader.ts

```diff
@@ -0,0 +1,122 @@
+    this.api = new Api({ rpc, signatureProvider, textDecoder: new TextDecoder(), textEncoder: new TextEncoder() })
+  }
+
+  public async getHeadBlockNumber(): Promise<number> {
+    const statusRow = await this.db.fill_status.findOne()
+    return statusRow.head
+  }
+
+  public async getLastIrreversibleBlockNumber(): Promise<number> {
+    const statusRow = await this.db.fill_status.findOne()
+    return statusRow.irreversible
+  }
+
+  public async getBlock(blockNumber: number): Promise<Block> {
+    const pgBlockInfo = await this.db.block_info.findOne({
+      block_index: blockNumber,
+    })
+
+    // Uses ${<var-name>} for JS substitutions and $<var-name> for massivejs substitutions.
+    const query = `
+      SELECT at.transaction_id, at.block_index, at.account,
+              at.name, at_authorization.actor, at_authorization.permission,
+              at.action_index, at.receipt_receiver, at.data
+      FROM ${this.dbSchema}.action_trace AS at, ${this.dbSchema}.action_trace_authorization AS at_authorization
+      WHERE at.block_index = $1 AND at.transaction_id = at_authorization.transaction_id
+    `
+
+    if (! this.massiveInstance) {
+      throw new NotInitializedError('Massive was not initialized.')
+    }
+
+    const pgActionTraceAuthorizations = await this.massiveInstance.query(query, [blockNumber])
+
+    const blockInfo = {
+      blockNumber: pgBlockInfo.block_index,
+      blockHash: pgBlockInfo.block_id,
+      previousBlockHash: pgBlockInfo.previous,
+      timestamp: pgBlockInfo.timestamp,
+    }
+
+    const actionPromises: Array<Promise<Action>> = pgActionTraceAuthorizations.map(async (actionTrace: any) => {
+      const serializedAction = {
+        account: actionTrace.account,
+        name: actionTrace.name,
+        authorization: [{
+          actor: actionTrace.actor,
+          permission: actionTrace.permission,
+        }],
+        data: actionTrace.data,
+      }
+
+      const [deserializedAction] = await this.deserializeActionTraces([serializedAction])
+
+      const action = {
+        type: `${actionTrace.account}::${actionTrace.name}`,
+        payload: {
+          ...deserializedAction,
+          actionIndex: actionTrace.action_index,
+          transactionId: actionTrace.transaction_id,
+          notifiedAccounts: [actionTrace.receipt_receiver]
+        }
+      }
+      return action
+    })
+
+    const actions = await Promise.all(actionPromises)
+
+    return {
+      blockInfo,
+      actions,
+    }
+  }
+
+  protected async setup(): Promise<void> {
+    if (this.initialized) {
+      return
+    }
+
+    try {
+      this.massiveInstance = await massive(this.massiveConfig)
+      this.db = this.massiveInstance[this.dbSchema]
+    } catch (err) {
+      throw new NotInitializedError('', err)
+    }
+  }
+
+  private async deserializeActionTraces(actionTraces: any): Promise<any> {
+    const actions = await this.api.deserializeActions(actionTraces)
```

Comment:
So, I was looking into if it was completely necessary to have an RPC endpoint, if the ABI information was already populated in Postgres. Todd confirmed with me that API information is stored in the `accounts` table. After looking at the `Api` object in eosjs, it seems that if you pass in your own [`AbiProvider`](https://github.com/EOSIO/eosjs/blob/master/src/eosjs-api.ts#L60), then we should be able to drop the requirement for an RPC endpoint completely. (Since it's required to instantiate an `Api` instance, we can just pass in something fake.) Here's the call stack up to `getRawAbi`, which you'd implement to pull from the `accounts` table instead:
```
Api#deserializeActions
  -> #getContract
    -> #getAbi
      -> #getCachedAbi
        -> #abiProvider.getRawAbi
```
As long as we're only calling `Api#deserializeActions`, we should have no problems.